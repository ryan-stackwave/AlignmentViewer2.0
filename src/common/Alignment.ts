import { generateUUIDv4 } from "./Utils";
import { defineNewAlignmentMode } from "./AceAlignmentMode";
import { Nucleotide } from "./Residues";
import { SequenceSorter } from "./AlignmentSorter";
import {
  AlignmentTypes,
  AminoAcidAlignmentStyle,
  NucleotideAlignmentStyle,
} from "./MolecularStyles";

export interface ISequence {
  id: string;
  sequence: string;
}

/**
 * Alignment
 * This class represents a multiple sequence alignment. During initialization,
 * positional statistics are generated about the alignment and a consensus
 * sequence is determined.
 *
 * The class should generally be instantiated from subclasses (e.g., FastaAlignemtn)
 * using static methods, e.g.,
 *     FastaAlignment.fromFileContents
 *     StockholmAlignment.fromFileContents
 */
export class Alignment {
  private uuid: string;
  private aceEditorMode: string | undefined;
  private name: string;
  private predictedNT: boolean;
  private sequences: Map<SequenceSorter, ISequence[]>;
  private querySequence: ISequence;
  private positionalLetterCounts = new Map<
    number,
    { [letter: string]: number }
  >();
  private globalAlphaLetterCounts: { [letter: string]: number } = {};
  private allUpperAlphaLettersInAlignmentSorted: string[];
  private consensus: {
    sequence: ISequence;
    statistics: {
      letter: string;
      position: number;
      occurrences: number;
    }[];
  };

  /**
   * Normalize all values in an object that contains all counts such that
   * the sum of those counts equals 1.
   * @param letterCounts an object that contains the same keys, but whose
   *                     values now sum to 1.
   * @param validLetters if provided, returned object will only contain letters
   *                     that are in this array.
   */
  private static normalizeLetterCounts(
    letterCounts: { [letter: string]: number },
    validLetters?: string[]
  ) {
    const totalCounts = Object.keys(letterCounts).reduce((acc, letter) => {
      if (!validLetters || validLetters.includes(letter)) {
        acc += letterCounts[letter];
      }
      return acc;
    }, 0);

    return Object.keys(letterCounts).reduce((acc, letter) => {
      if (!validLetters || validLetters.includes(letter)) {
        acc[letter] = letterCounts[letter] / totalCounts;
      }
      return acc;
    }, {} as { [letter: string]: number });
  }

  /**
   * Create a new Alignment object. The default query sequence will
   * be set to the first sequence in sequences. This can be changed
   * by calling setQuerySequence() after creation
   * @param sequences
   */
  public constructor(name: string, sequencesAsInput: ISequence[]) {
    this.uuid = generateUUIDv4();
    this.name = name;
    this.sequences = new Map<SequenceSorter, ISequence[]>();
    this.sequences.set(SequenceSorter.INPUT, sequencesAsInput);
    this.querySequence = sequencesAsInput[0];
    this.predictedNT = true;

    //
    //generate statistics
    //
    const start = new Date();
    const allLettersInAlignment: { [key: string]: number } = {}; //all letters in the alignment

    // aggregate stats for each position and globally including
    // the number of time each amino acid occurs
    // ** this loop takes the bulk of the initialization time
    for (
      let sequenceIdx = 0;
      sequenceIdx < sequencesAsInput.length;
      sequenceIdx++
    ) {
      const seq = sequencesAsInput[sequenceIdx].sequence;
      for (let positionIdx = 0; positionIdx < seq.length; positionIdx++) {
        const letter = seq[positionIdx];
        const letterIsAlpha = letter.match(/[a-z]/i) ? true : false;
        if (
          letterIsAlpha &&
          this.predictedNT &&
          Nucleotide.fromSingleLetterCode(letter) === Nucleotide.UNKNOWN
        ) {
          this.predictedNT = false;
        }

        const position = positionIdx; // zero based
        allLettersInAlignment[letter] = 1;

        let letterCounts = this.positionalLetterCounts.get(position);
        if (!letterCounts) {
          letterCounts = {};
          this.positionalLetterCounts.set(position, letterCounts);
        }
        if (letter in letterCounts === false) {
          letterCounts[letter] = 0;
        }
        letterCounts[letter] += 1;

        if (letterIsAlpha) {
          if (letter in this.globalAlphaLetterCounts === false) {
            this.globalAlphaLetterCounts[letter] = 0;
          }
          this.globalAlphaLetterCounts[letter] += 1;
        }
      }
    }

    //get a list of all the upper letters found in the alignment
    this.allUpperAlphaLettersInAlignmentSorted = Object.keys(
      allLettersInAlignment
    )
      .sort()
      .reduce((arr: string[], value: string) => {
        if (value.match(/[A-Z]/)) {
          // only keep letters
          arr.push(value);
        }
        return arr;
      }, []);

    //determine consensus sequence
    const consensusStats = Array.from(this.positionalLetterCounts).map(
      ([position, letterCounts]) => {
        const topLetter = Object.keys(letterCounts).reduce((a, b) => {
          const aIsLetter = a.match(/[a-z]/i);
          const bIsLetter = b.match(/[a-z]/i);
          if (aIsLetter && !bIsLetter) {
            return a;
          }
          if (bIsLetter && !aIsLetter) {
            return b;
          }
          return letterCounts[a] > letterCounts[b] ? a : b;
        });
        return {
          position: position,
          letter: topLetter,
          occurrences: letterCounts[topLetter],
        };
      }
    );
    this.consensus = {
      sequence: {
        id: "consensus",
        sequence: consensusStats.map((s) => s.letter).join(""),
      },
      statistics: consensusStats,
    };

    console.log(
      "done parsing alignment. took " +
        (new Date().getTime() - start.getTime()) +
        "ms"
    );
  }

  /**
   * Is this alignment predicted to be nucleotides?
   * @returns true if it is predicted to be nucleotide sequences.
   */
  getPredictedType(): AlignmentTypes {
    return this.predictedNT
      ? AlignmentTypes.NUCLEOTIDE
      : AlignmentTypes.AMINOACID;
  }

  /**
   * Returns a default nucleotide or amino acid style for this alignment,
   * predicted from the sequence itself.
   */
  getDefaultStyle(): AminoAcidAlignmentStyle | NucleotideAlignmentStyle {
    return this.predictedNT
      ? new NucleotideAlignmentStyle()
      : new AminoAcidAlignmentStyle();
  }

  /**
   * Get a unique uuid for this alignment.
   * @returns a uuid string for this alignment
   */
  getUUID() {
    return this.uuid;
  }

  /**
   * Get the name of this alignment
   * @returns a name for the alignment - usually the filename
   */
  getName() {
    return this.name;
  }

  /**
   * Get an ace editor mode for this alignemnt. If one doesn't exist
   * then it will be defined.
   * @returns the ace editor mode for this alignment
   */
  getAceEditorMode() {
    if (!this.aceEditorMode) {
      this.aceEditorMode = "ace/mode/" + this.uuid;
      defineNewAlignmentMode(this.aceEditorMode, this);
    }
    return this.aceEditorMode;
  }

  /**
   * Get the consensus sequence
   * @returns a position ordered array of objects that contain each letter, its number of
   *          occurrences, and the actual position number
   */
  getConsensus() {
    return this.consensus;
  }

  /**
   * Get the number of gaps at each position
   * @returns an array of gap counts, ordered by position
   */
  getGapsPerColumn(): number[] {
    return Array.from(this.positionalLetterCounts).map(
      ([position, letterCounts]) => {
        return !isNaN(letterCounts["-"]) ? letterCounts["-"] : 0;
      }
    );
  }

  /**
   * Get the number of gaps at a specific position
   * @returns the gap counts at the specified position
   */
  getGapCountAtColumn(position: number): number {
    const letterCounts = this.positionalLetterCounts.get(position);
    if (!letterCounts || !letterCounts["-"]) {
      return 0;
    }
    return letterCounts["-"];
  }

  /**
   * Get the total number of times each letter appears in the alignment.
   * @param normalize if true, will normalize the returned array such that
   *                  it sums to 1.
   * @param validLetters if provided, only those letters will be included
   *                     in the resulting counts
   */
  getGlobalAlphaLetterCounts(normalize?: boolean, validLetters?: string[]) {
    if (normalize) {
      return Alignment.normalizeLetterCounts(
        this.globalAlphaLetterCounts,
        validLetters
      );
    }
    return !validLetters
      ? this.globalAlphaLetterCounts
      : Object.keys(this.globalAlphaLetterCounts).reduce((acc, letter) => {
          //don't normalize, but remove invalid letters
          if (validLetters.includes(letter)) {
            acc[letter] = this.globalAlphaLetterCounts[letter];
          }
          return acc;
        }, {} as { [letter: string]: number });
  }

  /**
   * Get the length of the longest sequence
   * @returns the length of the largest sequence in this alignment
   */
  getMaxSequenceLength(): number {
    return this.getSequences().reduce((acc, seq) => {
      if (seq.sequence.length > acc) {
        acc = seq.sequence.length;
      }
      return acc;
    }, -1);
  }

  /**
   * Get the frequencies of all letters in this alignment for each position.
   * @param normalize if true, will normalize the returned letter counts for each
   *                  position such that each sums to 1.
   * @param validLetters if provided, only those letters will be included
   *                     in the resulting counts
   * @returns a Map whose keys are position (ordered) and values are a dictionary
   *          with [key = letter (e.g., amino acid code)] and [value = # occurrences].
   * TODO: Test parameter use cases.
   */
  getPositionalLetterCounts(normalize?: boolean, validLetters?: string[]) {
    if (normalize || validLetters) {
      return Array.from(this.positionalLetterCounts).reduce(
        (acc, [position, letterCounts]) => {
          acc.set(
            position,
            normalize
              ? Alignment.normalizeLetterCounts(letterCounts, validLetters) // normalize + remove invalid letters
              : Object.keys(letterCounts).reduce((acc2, letter) => {
                  //don't normalize, but remove invalid letters
                  if (validLetters!.includes(letter)) {
                    acc2[letter] = letterCounts[letter];
                  }
                  return acc2;
                }, {} as { [letter: string]: number })
          );
          return acc;
        },
        new Map<Number, { [letter: string]: number }>()
      );
    }
    return this.positionalLetterCounts;
  }

  /**
   * Get all the sequences in this alignment - ordered as they were input
   * @param sortBy specifies the order of the returned sequences. If it isn't
   *               provided, returns the sequences as they were input to create
   *               the Alignment object, e.g., usually fasta order.
   * @param forceReSort if provided, sequences are forced to be re-sorted (rather
   *                    than possibly cached)
   * @returns all sequences in this alignment
   */
  getSequences(sortBy?: SequenceSorter, forceReSort?: boolean): ISequence[] {
    sortBy = sortBy ? sortBy : SequenceSorter.INPUT;
    forceReSort = forceReSort ? forceReSort : false;
    if (
      sortBy !== SequenceSorter.INPUT &&
      (!this.sequences.has(sortBy) || forceReSort)
    ) {
      //cache not yet populated for this sort option
      this.sequences.set(
        sortBy,
        sortBy.sortFn(this.sequences.get(SequenceSorter.INPUT)!, this)
      );
    }
    return this.sequences.get(sortBy)!;
  }

  /**
   * Get a sorted list of all of the upper case letters in the alignment.
   * @returns a list of all letters in the alignment sorted alphabetically.
   */
  getAllUpperAlphaLettersInAlignmentSorted(): string[] {
    return this.allUpperAlphaLettersInAlignmentSorted;
  }

  /**
   * Get all the first sequence in the alignment, aka, target / query / first
   * @returns all sequences in this alignment
   */
  getQuerySequence(): ISequence {
    return this.querySequence;
  }

  /**
   * Set the query sequence
   * @param newQuery The new query sequence. Must already exist in the alignment.
   * @throws an error if the query sequence does not yet exist in the alignment
   *
   * IMPORTANT: Requires re-sorting
   */
  /*setQuerySequence(newQuery: ISequence) {
    const queryInAlignment = this.getSequences().find((seq) => {
      if (seq === newQuery) {
        return true;
      }
      return false;
    });
    if (queryInAlignment !== undefined) {
      this.querySequence = newQuery;

      //the resort could be really slow ... need to rethink.
      SequenceSorter.list.forEach((sortOption) => {
        //updated cached sorted sequences
        if (sortOption.isQuerySequenceDependent) {
          this.getSequences(sortOption, true);
        }
      });
      this.sequences.forEach(
        (sequences: ISequence[], sortBy: SequenceSorter) => {
          this.getSequences(sortBy, true); // force re-sorting, which updates cache
        }
      );
    } else {
      throw Error("query sequence does not exist in alignment");
    }
  }*/
}
