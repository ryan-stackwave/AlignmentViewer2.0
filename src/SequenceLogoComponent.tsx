/**
 *
 * Sequence logo component.
 * Inspired / derived from https://github.com/weng-lab/logojs-package
 *  (but much simpler)
 */

import React from "react";
import Alignment from "./Alignment";
import { GlyphFactory } from "./SequenceLogoGlyphs";
import { acePrefix } from "./MolecularStyles";

export enum LOGO_TYPES {
  LETTERS = "letter stack",
  BARS = "bar plot"
}

interface IGlyphFrequency {
  frequency: number;
  letter: {
    letter: string;
    classNames: string;
  };
}
interface IGlyphStack extends Array<IGlyphFrequency> {}

export interface ISequenceLogoComponentProps {
  alignment: Alignment;
  glyphWidth: number;
  logoType: LOGO_TYPES;

  logoLoaded(node: HTMLDivElement): void;

  readonly id: string;
}

export class SequenceLogoComponent extends React.Component<
  ISequenceLogoComponentProps
> {
  private scrollerDiv?: HTMLDivElement | null;
  private loaded: boolean = false;

  divLoaded(e: HTMLDivElement | null) {
    if (e && !this.scrollerDiv) {
      //only run if first time element available
      this.scrollerDiv = e;
      this.props.logoLoaded(this.scrollerDiv);
    }
  }

  shouldComponentUpdate(nextProps: ISequenceLogoComponentProps) {
    if (
      this.loaded === false ||
      this.props.logoType !== nextProps.logoType ||
      this.props.glyphWidth !== nextProps.glyphWidth
    ) {
      return true;
    }
    return false;
  }

  render() {
    if (!this.props.alignment || !this.props.glyphWidth) {
      return null;
    }

    const positionalLetterCounts = this.props.alignment.getPositionalLetterCounts();
    const lettersSorted = this.props.alignment.getSortedAlphaLetters();

    //calculate the frequencies of all letters in each column
    const frequencies: number[][] = [];
    positionalLetterCounts.forEach((letterCounts, position) => {
      const totalCountAtPosition = Object.keys(letterCounts).reduce(
        (total, letter) => {
          total += letterCounts[letter];
          return total;
        },
        0
      );

      const frequenciesAtPosition: number[] = lettersSorted.map(letter => {
        let freq = 0;
        if (letter in letterCounts) {
          freq = letterCounts[letter] / totalCountAtPosition;
        }
        return freq;
      });

      frequencies.push(frequenciesAtPosition);
    });

    //load alphabet information
    const sortedAlphabet = lettersSorted.map(letter => {
      return {
        letter: letter,
        classNames: acePrefix + letter
      };
    });

    //munge in alphabet and sort by frequency
    const logoData = frequencies.map((frequenciesAtPosition, positionIdx) => {
      return frequenciesAtPosition
        .map((frequency, letterIdx) => {
          return {
            frequency: frequency,
            letter: sortedAlphabet[letterIdx]
          };
        })
        .filter(freq => {
          return freq.frequency !== 0;
        })
        .sort((a, b) => (a.frequency > b.frequency ? 1 : -1));
    });

    const totalWidth = frequencies.length * this.props.glyphWidth;
    const totalHeight = 100;
    this.loaded = true;

    /**
     * Generate the svg elements for a single position, i.e., column
     *
     */
    const generateStack = (positionalFrequencies: IGlyphStack) => {
      let dy = 100;

      const xscale = 2 / logoData.length; // not exact, but works.
      return positionalFrequencies.map((freq, idx) => {
        dy = dy - freq.frequency * 100;

        if (this.props.logoType === LOGO_TYPES.BARS) {
          return (
            <rect
              width="100"
              height="100"
              transform={`translate(0, ${dy}) scale(${xscale},${freq.frequency})`}
              className={freq.letter.classNames}
              key={`idxrect_${idx}`}
            ></rect>
          );
        }

        const selectedGlyph = GlyphFactory.glyphFromChar(freq.letter.letter)({
          className: freq.letter.classNames,
          transform: `translate(0, ${dy}) scale(${xscale},${freq.frequency})`,
          key: `idxglyph_${idx}`
        });
        return selectedGlyph;
      });
    };

    return (
      <div id={this.props.id} ref={e => this.divLoaded(e)}>
        <svg
          preserveAspectRatio="none"
          viewBox={`0 0 ${logoData.length} 100`}
          style={{ width: totalWidth, height: totalHeight }}
          xmlns="http://www.w3.org/2000/svg"
        >
          {logoData.map((singlePositionData, positionIdx) => {
            return (
              <g
                transform={`translate(${positionIdx},0)`}
                key={"p_" + positionIdx}
              >
                {generateStack(singlePositionData)}
              </g>
            );
          })}
        </svg>
      </div>
    );
  }
}
