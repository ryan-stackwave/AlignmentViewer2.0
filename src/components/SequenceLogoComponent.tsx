/**
 * Sequence logo component.
 * Inspired / derived from https://github.com/weng-lab/logojs-package
 *  (but simpler)
 */

import React from "react";
import "./SequenceLogo.scss";
import { Alignment } from "../common/Alignment";
import { GlyphFactory } from "../common/SequenceLogoGlyphs";
import {
  aceResidueParentClass,
  getLetterClassNames,
  AlignmentTypes,
} from "../common/MolecularStyles";
import ReactTooltip from "react-tooltip";
import { AminoAcid, Nucleotide } from "../common/Residues";

export enum LOGO_TYPES {
  LETTERS = "Letter Stack",
  BARS = "Bar Plot",
}

interface ILetterWithClasses {
  letter: string;
  classNames: string;
}
interface IGlyphFrequency {
  count: number;
  frequency: number;
  letter: ILetterWithClasses;
  residue: AminoAcid | Nucleotide;
}
interface IGlyphStackData extends Array<IGlyphFrequency> {}

export interface ISequenceLogoProps {
  alignment: Alignment;
  glyphWidth: number;
  height: number;
  logoType: LOGO_TYPES;
  alignmentType?: AlignmentTypes;
}

interface ISequenceLogoState {
  logoData?: IGlyphStackData;
}

export class SequenceLogoComponent extends React.Component<
  ISequenceLogoProps,
  ISequenceLogoState
> {
  private logoData?: IGlyphStackData[];

  static defaultProps = {
    logoType: LOGO_TYPES.LETTERS,
    height: 100,
  };

  /*
   *
   *
   * PRIVATE METHODS
   *
   *
   */

  private renderTooltip() {
    return (
      <ReactTooltip
        id="getLogoTooltip"
        effect="solid"
        type="light"
        border={true}
        getContent={(idx: string) => {
          if (!idx || !this.logoData) {
            return;
          }

          const position = parseInt(idx) + 1;
          const glyphData = this.logoData![parseInt(idx)];
          return glyphData.length === 0 ? null : (
            <div className="logo-tooltip">
              <h1>Position: {position}</h1>
              <div className="row">
                <div className="col legend-square"></div>
                <div className="col header name">Residue</div>
                <div className="col header frequency">Frequency</div>
                <div className="col header count">Count</div>
              </div>
              {glyphData
                .slice()
                .reverse()
                .map((letterFreq, myidx) => {
                  return (
                    <div key={myidx} className="row">
                      <div
                        className={`col legend-square ${letterFreq.letter.classNames}`}
                      ></div>
                      <div className="col name">
                        {`${letterFreq.residue.fullName} [${letterFreq.letter.letter}]`}
                      </div>
                      <div className="col frequency">
                        {+(letterFreq.frequency * 100).toFixed(2)}%
                      </div>
                      <div className="col count">{letterFreq.count}</div>
                    </div>
                  );
                })}
            </div>
          );
        }}
      />
    );
  }

  /**
   * Generate the svg elements for a single position, i.e., column
   */
  private renderSinglePositionStack(
    positionalFrequencies: IGlyphStackData,
    numGlyphStacks: number
  ) {
    const { logoType, height } = this.props;
    let dy = height;

    return positionalFrequencies.map((freq, idx) => {
      dy = dy - freq.frequency * 100;

      if (logoType === LOGO_TYPES.BARS) {
        const xscale = 2 / numGlyphStacks; // not exact, but works.
        return (
          <rect
            width="100"
            height="100"
            transform={`translate(0, ${dy}) scale(${xscale},${freq.frequency})`}
            className={freq.letter.classNames}
            key={`idxrect_${idx}`}
          ></rect>
        );
      }

      const xscale = 0.009; // invariant scaling
      const selectedGlyph = GlyphFactory.glyphFromChar(freq.letter.letter)({
        className: freq.letter.classNames,
        transform: `translate(0, ${dy}) scale(${xscale},${freq.frequency})`,
        key: `idxglyph_${idx}`,
      });
      return selectedGlyph;
    });
  }

  /**
   * Munge letter count data that was calculated during alignment creation
   * into a form appropriate for the glyph generation
   */
  private mungeLogoData(): IGlyphStackData[] {
    const { alignment } = this.props;

    const numberSequences = alignment.getSequences().length;
    const lettersSorted = alignment.getAllUpperAlphaLettersInAlignmentSorted();

    const alignmentType = this.props.alignmentType
      ? this.props.alignmentType
      : this.props.alignment.getPredictedType();
    const moleculeClass =
      alignmentType === AlignmentTypes.AMINOACID ? AminoAcid : Nucleotide;

    //load class names for each letter
    const letterObjects = lettersSorted.reduce((arr, letter) => {
      arr[letter] = {
        letter: letter,
        classNames: getLetterClassNames(letter, false, false, false),
      };
      return arr;
    }, {} as { [letter: string]: ILetterWithClasses });

    //note: removes invalid letters, but letterCount (value) isn't sorted
    const plc = alignment.getPositionalLetterCounts(false, lettersSorted);
    return Array.from(plc).map(
      ([pos, letterCounts]): IGlyphStackData => {
        return Object.entries(letterCounts)
          .map(([letter, count]) => {
            return {
              count: count,
              frequency: count / numberSequences,
              letter: letterObjects[letter],
              residue: moleculeClass.fromSingleLetterCode(letter),
            };
          })
          .filter((freq) => {
            // remove empty letters
            return freq.frequency !== 0;
          })
          .sort((a, b) => (a.frequency > b.frequency ? 1 : -1));
      }
    );
  }

  /*
   *
   *
   * REACT METHODS
   *
   *
   */
  shouldComponentUpdate(nextProps: ISequenceLogoProps) {
    const { alignment, alignmentType, logoType, glyphWidth } = this.props;
    if (
      !this.logoData ||
      logoType !== nextProps.logoType ||
      glyphWidth !== nextProps.glyphWidth ||
      alignment !== nextProps.alignment ||
      alignmentType !== nextProps.alignmentType
    ) {
      return true;
    }
    return false;
  }

  render() {
    const { alignment, glyphWidth, height } = this.props;
    if (!alignment || !glyphWidth) {
      return null;
    }

    //perform a bunch of data munging
    this.logoData = this.mungeLogoData();
    const totalWidth = this.logoData.length * glyphWidth;
    const totalHeight = height;

    return (
      <div className="sequence-logo">
        <svg
          preserveAspectRatio="none"
          viewBox={`0 0 ${this.logoData.length} 100`}
          style={{ width: totalWidth, height: totalHeight }}
          xmlns="http://www.w3.org/2000/svg"
        >
          {this.logoData.map((singlePositionData, positionIdx) => {
            return (
              <g
                transform={`translate(${positionIdx},0)`}
                className={aceResidueParentClass} //required for default coloring
                key={"p_" + positionIdx}
              >
                {this.renderSinglePositionStack(
                  singlePositionData,
                  this.logoData!.length
                )}
                <rect
                  className="interaction-placeholder"
                  width="1"
                  height="100"
                  data-for="getLogoTooltip"
                  data-tip={positionIdx}
                  data-class={"sequence-logo-tooltip-container"}
                ></rect>
              </g>
            );
          })}
        </svg>
        {this.renderTooltip()}
      </div>
    );
  }
}
