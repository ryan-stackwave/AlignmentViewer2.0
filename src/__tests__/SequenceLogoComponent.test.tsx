import * as React from "react";
import { shallow, default as Enzyme } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import "jest-canvas-mock";

import Alignment from "../Alignment";
import { SequenceLogoComponent, LOGO_TYPES } from "../SequenceLogoComponent";

describe("SequenceLogoComponent", () => {
  beforeAll(() => {
    Enzyme.configure({ adapter: new Adapter() });
  });

  it("Should render when given default props.", () => {
    const wrapper = shallow(
      <SequenceLogoComponent
        alignment={new Alignment("", [])}
        glyphWidth={3}
        id={"test-sequence-logo-component"}
        logoType={LOGO_TYPES.BARS}
        logoLoaded={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("Should render when given a basic alignment.", () => {
    const wrapper = shallow(
      <SequenceLogoComponent
        alignment={
          new Alignment("my-alignment", [
            { id: "my-alignment-1", sequence: "CDE" },
            { id: "my-alignment-2", sequence: "EDC" }
          ])
        }
        glyphWidth={3}
        id={"test-sequence-logo-component"}
        logoType={LOGO_TYPES.BARS}
        logoLoaded={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
