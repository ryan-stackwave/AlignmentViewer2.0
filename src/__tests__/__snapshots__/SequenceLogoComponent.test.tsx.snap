// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SequenceLogoComponent Should render when given a basic alignment. 1`] = `
<div
  id="test-sequence-logo-component"
>
  <svg
    preserveAspectRatio="none"
    style={
      Object {
        "height": 100,
        "width": 9,
      }
    }
    viewBox="0 0 3 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      key="p_0"
      transform="translate(0,0)"
    >
      <rect
        className="acePrefixE"
        height="100"
        key="idxrect_0"
        transform="translate(0, 50) scale(0.6666666666666666,0.5)"
        width="100"
      />
      <rect
        className="acePrefixC"
        height="100"
        key="idxrect_1"
        transform="translate(0, 0) scale(0.6666666666666666,0.5)"
        width="100"
      />
    </g>
    <g
      key="p_1"
      transform="translate(1,0)"
    >
      <rect
        className="acePrefixD"
        height="100"
        key="idxrect_0"
        transform="translate(0, 0) scale(0.6666666666666666,1)"
        width="100"
      />
    </g>
    <g
      key="p_2"
      transform="translate(2,0)"
    >
      <rect
        className="acePrefixE"
        height="100"
        key="idxrect_0"
        transform="translate(0, 50) scale(0.6666666666666666,0.5)"
        width="100"
      />
      <rect
        className="acePrefixC"
        height="100"
        key="idxrect_1"
        transform="translate(0, 0) scale(0.6666666666666666,0.5)"
        width="100"
      />
    </g>
  </svg>
</div>
`;

exports[`SequenceLogoComponent Should render when given default props. 1`] = `
<div
  id="test-sequence-logo-component"
>
  <svg
    preserveAspectRatio="none"
    style={
      Object {
        "height": 100,
        "width": 0,
      }
    }
    viewBox="0 0 0 100"
    xmlns="http://www.w3.org/2000/svg"
  />
</div>
`;
